oauthapplication/client.go

```diff
@@ -0,0 +1,116 @@
+
+const path = "/oauth_applications"
+
+type Client struct {
+	Backend clerk.Backend
+}
+
+func NewClient(config *clerk.ClientConfig) *Client {
+	return &Client{
+		Backend: clerk.NewBackend(&config.BackendConfig),
+	}
+}
+
+// fetches a single OAuth application by its ID.
+func (c *Client) Get(ctx context.Context, id string) (*clerk.OAuthApplication, error) {
+	path, err := clerk.JoinPath(path, id)
+	if err != nil {
+		return nil, err
+	}
+	req := clerk.NewAPIRequest(http.MethodGet, path)
+	resource := &clerk.OAuthApplication{}
+	err = c.Backend.Call(ctx, req, resource)
+	return resource, err
+}
+
+type ListParams struct {
+	clerk.APIParams
+	clerk.ListParams
+}
+
+func (params *ListParams) ToQuery() url.Values {
+	q := params.ListParams.ToQuery()
+	return q
+}
+
+// Retrieves all OAuth applications.
+func (c *Client) List(ctx context.Context, params *ListParams) (*clerk.OAuthApplicationList, error) {
+	req := clerk.NewAPIRequest(http.MethodGet, path)
+	req.SetParams(params)
+	list := &clerk.OAuthApplicationList{}
+	err := c.Backend.Call(ctx, req, list)
+	return list, err
+}
+
+type OAuthApplicationCreateParams struct {
+	clerk.APIParams
+	Name        string `json:"name"`
+	CallbackURL string `json:"callback_url"`
+	Scopes      string `json:"scopes"`
+	Public      bool   `json:"public"`
+}
+
+// Creates a new OAuth application with the given parameters.
+func (c *Client) Create(ctx context.Context, params *OAuthApplicationCreateParams) (*clerk.OAuthApplicationWithClientSecret, error) {
+	req := clerk.NewAPIRequest(http.MethodPost, path)
+	req.SetParams(params)
+	authApplication := &clerk.OAuthApplicationWithClientSecret{}
+	err := c.Backend.Call(ctx, req, authApplication)
+	return authApplication, err
+}
+
+type OAuthApplicationUpdateParams struct {
+	clerk.APIParams
+	Name        *string `json:"name,omitempty"`
+	CallbackURL *string `json:"callback_url,omitempty"`
+	Scopes      *string `json:"scopes,omitempty"`
+}
+
+// Updates an existing OAuth application.
+func (c *Client) Update(ctx context.Context, id string, params *OAuthApplicationUpdateParams) (*clerk.OAuthApplication, error) {
+	path, err := clerk.JoinPath(path, id)
+	if err != nil {
+		return nil, err
+	}
+	req := clerk.NewAPIRequest(http.MethodPatch, path)
+	req.SetParams(params)
+	authApplication := &clerk.OAuthApplication{}
+	err = c.Backend.Call(ctx, req, authApplication)
+	return authApplication, err
+}
+
+// Deletes the given OAuth application
+func (c *Client) DeleteOAuthApplication(ctx context.Context, id string) (*clerk.DeletedResource, error) {
+	path, err := clerk.JoinPath(path, id)
+	if err != nil {
+		return nil, err
+	}
+	req := clerk.NewAPIRequest(http.MethodDelete, path)
+	authApplication := &clerk.DeletedResource{}
+	err = c.Backend.Call(ctx, req, authApplication)
+	return authApplication, err
+}
+
+// Rotates the OAuth application's client secret
+func (c *Client) RotateClientSecret(ctx context.Context, id string) (*clerk.OAuthApplicationWithClientSecret, error) {
```

Comment:
Let's use one type for `OAuthApplication`s.
```suggestion
func (c *Client) RotateClientSecret(ctx context.Context, id string) (*clerk.OAuthApplication, error) {
```